### PR TITLE
Add list optional T constructor for ArrowFixedWidthVector and change to dtypepointer

### LINKED ISSUE #9: Allow null values in ArrowFixedWidthVector
Right now it only allows non-null values. Create another constructor with the signature
```
fn __init__(inout self, values: List[Optional[T]]): ...
```

### PR DESCRIPTION
**Pull request by martinvuyk**
Closes #9

### CODE REVIEW

**Comment by sa-** [comment]
This is a nice change, could you please add tests for it?

**Comment by martinvuyk** [comment]
added a thorough test and changed this API a bit WDYT ?
```mojo
    fn __getitem__(self, index: Int) -> Optional[Scalar[T]]:
        if not (0 <= index < self.length):
            return None
        if not self.validity[index]:
            return None
        return self.view[index]
```

File: arrow/physical_layout/arrow.mojo [hunk 1]
```diff
@@ -5,44 +5,88 @@ from arrow.buffer.bitmap import Bitmap
...
         var num_bytes_with_padding = get_num_bytes_with_padding(num_bytes)
-        var ui8_ptr = Pointer[UInt8].alloc(
+        var ui8_ptr = Self._ptr_type.alloc(
             num_bytes_with_padding, alignment=ALIGNMENT
         )
         memset_zero(ui8_ptr, num_bytes_with_padding)
         var ptr = ui8_ptr.bitcast[T]()
 
-        var validity_list = List[Bool](len(values))
+        var validity_list = List[Bool](capacity=len(values))
```
Comment (sa-) [inline arrow/physical_layout/arrow.mojo:28]: This bug has bit me a few times, thanks for fixing it

File: arrow/physical_layout/arrow.mojo [hunk 2]
```diff
@@ -5,44 +5,88 @@ from arrow.buffer.bitmap import Bitmap
...
+        self.value = ui8_ptr
+        self.validity = Bitmap(validity_list)
+        self.null_count = null_count
+        self.view = ptr
+        self.length = len(values)
+        self.mem_used = num_bytes_with_padding
+
+    fn __getitem__(self, index: Int) -> Optional[Scalar[T]]:
+        if not (0 <= index < self.length):
+            return None
```
Comment (sa-) [inline arrow/physical_layout/arrow.mojo:75]: The change to account for optional values is nice but this check should still throw an out of bounds error. Not throwing an error here means that it's possible to have silent errors or unexpected behavior e.g. assuming `arr[-1]` would return the last item. The user mistakenly assumes it to be a valid and null value but it should be out of bounds.

**Review by sa- — CHANGES_REQUESTED** [review]

**Comment by sa-** [comment]
Did the tests pass on your machine? Were you using a nightly?

File: arrow/physical_layout/arrow.mojo (hunk 2 above)
Comment (martinvuyk) [reply arrow/physical_layout/arrow.mojo:75]: > Not throwing an error here means that it's possible to have silent errors or unexpected behavior e.g. assuming arr[-1] would return the last item.

that is what the Optional type is for, it has to be checked with bool(opt) before calling `.value()` on it

> The user mistakenly assumes it to be a valid and null value but it should be out of bounds.

they can't since it also checks the validity bitmap and any index could return None since it supports null values now.

**Comment by martinvuyk** [comment]
> Did the tests pass on your machine? Were you using a nightly?

yes they did then and they do now as well with nightly `mojo 2024.7.205`

File: arrow/buffer/bitmap.mojo [hunk 3]
```diff
@@ -78,9 +78,19 @@ struct Bitmap(StringableRaising):
...
+
+        Args:
+            index: The index [0, len(self) -1].
+
+        Returns:
+            The value.
+        """
+
+        if not (0 <= index < self.length):
+            return False
```
Comment (sa-) [inline arrow/buffer/bitmap.mojo:93]: Please change this back, a design change like this should be submitted as a separate PR instead of being added to this one!

File: arrow/physical_layout/arrow.mojo (hunk 2 above)
Comment (sa-) [reply arrow/physical_layout/arrow.mojo:85]: I can see where you are coming from, however at this point in the codebase we throw errors for out of bounds access instead of silent failures. To make the design change of returning values instead of throwing errors another separate pull request could be made. This pull request will only be merged if an error is thrown on out of bounds access
